File: samples/adapters/in_memory.js
/**
 * InMemoryClient implementation
 *
 * Implements the InMemoryClient Required Port.
 * Manages in-memory key-value storage for the current process.
 */

class InMemoryAdapter {
  constructor() {
    this.data = new Map();
  }

  /**
   * Get value from in memory
   * @param {string} key - The key to retrieve
   * @returns {any|null} The value or null if not found
   */
  get(key) {
    return this.data.get(key) ?? null;
  }

  /**
   * Set value in in memory
   * @param {string} key - The key to set
   * @param {any} value - The value to store
   */
  set(key, value) {
    this.data.set(key, value);
  }

  /**
   * Delete value from in memory
   * @param {string} key - The key to delete
   * @returns {boolean} True if deleted, false if not found
   */
  delete(key) {
    return this.data.delete(key);
  }

  /**
   * Clear all data
   */
  clear() {
    this.data.clear();
  }

  /**
   * Get all keys
   * @returns {string[]} Array of all keys
   */
  keys() {
    return Array.from(this.data.keys());
  }
}

module.exports = InMemoryAdapter;

File: samples/adapters/in_memory.rs
/// InMemoryClient implementation
///
/// Implements the InMemoryClient Required Port.
/// Manages in-memory key-value storage for the current process.

use serde_json::Value;
use std::collections::HashMap;
use state_engine::ports::required::InMemoryClient;

pub struct InMemoryAdapter {
    data: HashMap<String, Value>,
}

impl InMemoryAdapter {
    pub fn new() -> Self {
        Self {
            data: HashMap::new(),
        }
    }

    /// Clear all data
    pub fn clear(&mut self) {
        self.data.clear();
    }

    /// Get all keys
    pub fn keys(&self) -> Vec<String> {
        self.data.keys().cloned().collect()
    }
}

impl InMemoryClient for InMemoryAdapter {
    fn get(&self, key: &str) -> Option<Value> {
        self.data.get(key).cloned()
    }

    fn set(&mut self, key: &str, value: Value) {
        self.data.insert(key.to_string(), value);
    }

    fn delete(&mut self, key: &str) -> bool {
        self.data.remove(key).is_some()
    }
}
